alpenhorn/util.py

```diff
@@ -15,36 +15,51 @@
 log = logging.getLogger(__name__)
 
 
-def run_command(cmd, **kwargs):
+def run_command(
+    cmd: list[str], timeout: float | None = None, **kwargs
+) -> tuple(int | None, str, str):
     """Run a command.
 
     Parameters
     ----------
-    cmd : array
+    cmd : list of strings
         A command as a list of strings including all arguments.
-    kwargs : dict
-        Passed directly onto `subprocess.Popen.`
+    timeout : float or None
+        Number of seconds to wait before forceably killing the process,
+        or None to wait forever.
+
+    Other keyword args are passed directly on to subprocess.Popen
 
     Returns
     -------
-    retval : int
-        Return code.
-    stdout_val : string
-        Value of stdout.
-    stderr_val : string
-        Value of stderr.
+    Returns a three-element tuple containing:
+        retval : int or None
+            Return code, or None if the process was killed after timing out.
+        stdout : string
+            Value of stdout.
+        stderr : string
+            Value of stderr.
     """
 
     import subprocess
 
-    log.debug('Running command "%s"', " ".join(cmd))
+    log.debug(f"Running command [timeout={timeout}]: " + " ".join(cmd))
 
     # run using Popen
     proc = subprocess.Popen(
         cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, **kwargs
     )
-    stdout_val, stderr_val = proc.communicate()
-    retval = proc.returncode
+    if timeout is None:
```

Comment:
I don't think we actually have to check for `timeout` being `None` ourselves, the default value for `timeout` in `.communicate` is `None` so we can just pass that along and the behaviour won't change. (i.e., just keep the `try: except:` block)

Reply:
Oooo, nice.